homeassistant/components/stt/__init__.py

```diff
@@ -36,33 +45,166 @@
 
 __all__ = [
     "async_get_provider",
+    "async_get_speech_to_text_entity",
     "AudioBitRates",
     "AudioChannels",
     "AudioCodecs",
     "AudioFormats",
     "AudioSampleRates",
     "DOMAIN",
     "Provider",
+    "SpeechToTextEntity",
     "SpeechMetadata",
     "SpeechResult",
     "SpeechResultState",
 ]
 
+_LOGGER = logging.getLogger(__name__)
+
+
+@callback
+def async_get_speech_to_text_entity(
+    hass: HomeAssistant, entity_id: str
+) -> SpeechToTextEntity | None:
+    """Return stt entity."""
+    component: EntityComponent[SpeechToTextEntity] = hass.data[DOMAIN]
+
+    return component.get_entity(entity_id)
+
 
 async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up STT."""
+    component = hass.data[DOMAIN] = EntityComponent[SpeechToTextEntity](
+        _LOGGER, DOMAIN, hass
+    )
+
+    component.register_shutdown()
     platform_setups = async_setup_legacy(hass, config)
 
     if platform_setups:
         await asyncio.wait([asyncio.create_task(setup) for setup in platform_setups])
 
-    hass.http.register_view(SpeechToTextView(hass.data[DOMAIN]))
+    hass.http.register_view(SpeechToTextView(hass.data[DATA_PROVIDERS]))
     return True
 
 
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[SpeechToTextEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[SpeechToTextEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+class SpeechToTextEntity(RestoreEntity):
+    """Represent a single STT provider."""
+
+    _attr_should_poll = False
+    __last_processed: str | None = None
+
+    @property
+    @final
+    def name(self) -> str:
+        """Return the name of the provider entity."""
+        # Only one entity is allowed per platform for now.
+        if self.platform is None:
+            raise RuntimeError("Entity is not added to hass yet.")
+
+        return self.platform.platform_name
```

Comment:
What will this end up returning?

Reply:
The config entry domain.

Reply:
In a follow up I suggest to build the helper we talked about to lookup the correct engine based on either entity id or legacy provider name. We use that helper in the view api, and then we can allow the platforms to set any entity name they want here.